torchmetrics/classification/confusion_matrix.py

```diff
@@ -118,7 +118,10 @@ def __init__(
         if self.normalize not in allowed_normalize:
             raise ValueError(f"Argument average needs to one of the following: {allowed_normalize}")
 
-        default = torch.zeros(num_classes, 2, 2) if multilabel else torch.zeros(num_classes, num_classes)
+        if multilabel:
+            default = torch.zeros(num_classes, 2, 2, dtype=torch.long)
+        else:
+            default = torch.zeros(num_classes, num_classes, dtype=torch.long)
```

Comment:
This won't be sufficient. We also need to make sure that the dtype is also preserved when calling metric.to(`torch.float`). IMO in that case we should keep the internal state like that and only convert the output dtype

Reply:
@justusschock Actually, this will work since PR https://github.com/PyTorchLightning/metrics/pull/493 made sure that `metric.to(torch.float)` will no longer change the underlying data type of the metric state (we saw problems with mixed precision). User have to use `metric.set_dtype(torch.float)` to change the dtype and in that case its on the user that the change the precision.

Reply:
but still some tests are failing :]

Reply:
fixed :rabbit: